sqlite_data_norm.c

```diff
@@ -142,21 +143,68 @@ sqlite_fdw_uuid_blob (const unsigned char* s0, unsigned char* Blob)
  */
 
 /*
- * uuid_str converts a UUID X into a well-formed UUID string.
+ * aBlob to RFC UUID string with 36 characters
+ */
+
+static void
+sqlite3UuidBlobToStr( const unsigned char *aBlob, unsigned char *zs)
+{
+	static const char hex_dig[] = "0123456789abcdef";
+	int i, k;
+	unsigned char x;
+	k = 0;
+	for(i=0, k=0x550; i<UUID_LEN; i++, k=k>>1)
+	{
+		if( k&1 )
+		{
+			zs[0] = '-';
+			zs++;
+		}
+		x = aBlob[i];
+		zs[0] = hex_dig[x>>4];
+		zs[1] = hex_dig[x&0xf];
+		zs += 2;
+	}
+	*zs = 0;
+}
+
+/*
+ * Converts argument BLOB-UUID into a well-formed UUID string.
  * X can be either a string or a blob.
- *
- * static void uuid_str(sqlite3_context* context, int argc, sqlite3_value** argv) {
- *	unsigned char aBlob[16];
- *	unsigned char zs[37];
- *	const unsigned char* pBlob;
- *	(void)argc;
- *	pBlob = sqlite_fdw_data_norm_uuid_input_to_blob(argv[0], aBlob);
- *	if (pBlob == 0)
- *		return;
- *	sqlite_fdw_data_norm_uuid_blob_to_str(pBlob, zs);
- *	sqlite3_result_text(context, (char*)zs, 36, SQLITE_TRANSIENT);
- *}
  */
+static void
+sqlite_fdw_uuid_str(sqlite3_context* context, int argc, sqlite3_value** argv)
+{
+	unsigned char aBlob[UUID_LEN];
+	const unsigned char* pBlob;
+	unsigned char zs[UUID_LEN * 2 + 1];
+	sqlite3_value* arg = argv[0];
+	int t = sqlite3_value_type(arg);
+
+	if (t == SQLITE_BLOB)
+	{
+		pBlob = sqlite3_value_blob(arg);
+	}
+	if (t == SQLITE3_TEXT)
+	{
+		const unsigned char* txt = sqlite3_value_text(arg);
+		if (sqlite_fdw_uuid_blob(txt, aBlob))
+			pBlob = aBlob;
+		else
+		{
+			sqlite3_result_null(context);
+			return;
+		}
+	}
+	if (t != SQLITE_BLOB)
```

Comment:
I think the logic is not correct for TEXT type.  
If `t == SQLITE3_TEXT`, I think the code flow will be like this.  
Firstly, it runs into the code block of `if (t == SQLITE3_TEXT)` (line 188) and set `pBlob = aBlob;` (line 192).  
Next, it runs into this block, and then return NULL immediately.  
Is that correct behavior?  
I think the correct source code should be like this.  
```
if (t == SQLITE_BLOB)   --> Line 184
...
else if (t == SQLITE3_TEXT)   --> Line 188
...
else   --> Line 199
{
   sqlite3_result_null(context);
   return;
}
```

Reply:
The logic was simplifyed, @son-phamngoc 
1. SQLITE_BLOB => use original BLOB
2. SQLITE3_TEXT => use BLOB from normalized string, if no normal string `NULL`
3. Otherwise `NULL`
4.  If "no `NULL`" branch, use BLOB (`sqlite3UuidBlobToStr(pBlob, zs);`)

This function was not intuitive. Is it better now?

Reply:
Thanks for fixing. It looks better. I confirmed. Please help me to resolve this comment.